File: datastructures/graphs/disjoint_sets.py
## Disjoint set (alias UnionFind) data structure implementation

class DisjointSet:
    def __init__(self, size):
        """
        Initialize root array to len `size` with the values
        acting as the roots and the indices as the node values.
        Initially the root of each node(index) is itself.
        """
        self.root_array = [i for i in range(size)]


    def find(self, x):
        """Find root of node `x`.
        """
        return self.root_array[x]


    def union(self, x, y):
        """Add new connections `x`,`y` to the disjoint set.
        """
        root_x = self.find(x)
        root_y = self.find(y)

        ## if root_x == root_y, then the nodes are already
        ## connected, else make traverse the root array and set
        ## the root (value at root[i]==root_y) of the index equal to
        ## the value of root_y to root_x
        if root_x != root_y:
            for i in range(len(self.root_array)):
                if self.root_array[i] == root_y:
                    self.root_array[i] = root_x


    def is_connected(self, x, y):
        """Check if nodes x and y are connected.
        Two nodes are connected if they have the same
        root node in the root array.
        """
        return self.find(x) == self.find(y)



if __name__=="__main__":
    d_set = DisjointSet(10)
    d_set.union(1, 2)
    d_set.union(2, 5)
    d_set.union(5, 6)
    d_set.union(6, 7)
    d_set.union(3, 8)
    d_set.union(8, 9)
    # graph: 1-2-5-6-7 3-8-9
    print(d_set.is_connected(1, 5))  # true
    print(d_set.is_connected(5, 7))  # true
    print(d_set.is_connected(4, 9))  # false
    #print(d_set.find(6))
    print(d_set.root_array)
    
    d_set.union(9, 4) # graph: 1-2-5-6-7 3-8-9-4
    print(d_set.is_connected(4, 9))  # true
    #print(d_set.find(9))

File: datastructures/graphs/disjoint_sets.cpp
// Disjoint set (alias UnionFind) data structure implementation
#include<iostream>
#include<vector>

//using namespace std;
class DisjointSet {

    std::vector<int> root_array;

    public:
        /* alternative intialization of root_array size
         DisjointSet(int size){
            root_array(size);
            ...
         }
        */
        DisjointSet(int size) : root_array(size) {
            // Constructor: initialize root of each node(index) to itself
            for (int i = 0; i < size; i++) {
                root_array[i] = i;
            }
        }

        int find(int x) {
            // find root of node x
            return root_array[x];
        }

        void unionSet(int x, int y) {
            // Add new node connections to the disjoint set
            int root_x = find(x);
            int root_y = find(y);

            if (root_x != root_y) {
                for (int i = 0; i < root_array.size(); i++) {
                    if (root_array[i] == root_y) {
                        root_array[i] = root_x;
                    }
                }
            }
        }

        bool is_connected(int x, int y) {
            // Check if two nodes are connected.
            // Two nodes are connected if they have the same root
            return find(x) == find(y);
        }

};


// Test 
int main() {
    DisjointSet d_set(10);
    d_set.unionSet(1, 2);
    d_set.unionSet(2, 5);
    d_set.unionSet(5, 6);
    d_set.unionSet(6, 7);
    d_set.unionSet(3, 8);
    d_set.unionSet(8, 9);
    // graph: 1-2-5-6-7 3-8-9
    std::cout << std::boolalpha; //display bool string literal instead of 0, 1
    std::cout << d_set.is_connected(1, 5) << std::endl; //true
    std::cout << d_set.is_connected(5, 7) << std::endl; //true
    std::cout << d_set.is_connected(4, 9) << std::endl; //false

    d_set.unionSet(9, 4); // graph: 1-2-5-6-7 3-8-9-4
    //std::cout << d_set.root_array << std::endl;
    std::cout << d_set.is_connected(4, 9) << std::endl; //true

    return 0;
}